### DISCUSSION TITLE
Using with Nextjs13 Server Component

### DISCUSSION THREAD

**Discussion by cqh963852** [saadeghi/daisyui#1759] [Q&A] [answered]
Nextjs 13 provider a feature `appDir`. This can make component work as server component.

I have create a demo page with drawer component.

While with bad network, the drawer will has a flash. What should I do?

https://user-images.githubusercontent.com/17702287/232524092-bf7dac9d-a1c0-44e3-a88c-991d5379ef26.mp4

---

update:

demo repo https://github.com/cqh963852/nextjs-13-daisyui

codesandbox https://codesandbox.io/p/sandbox/modest-forest-n477kw?file=%2Fapp%2Fglobals.css&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A4%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A4%7D%5D

code

```
export default function Home() {
  return (
    <div className="drawer">
      <input id="my-drawer" type="checkbox" className="drawer-toggle" />
      <div className="drawer-content">
        <label htmlFor="my-drawer" className="btn btn-primary drawer-button">
          Open drawer
        </label>
      </div>
      <div className="drawer-side">
        <label htmlFor="my-drawer" className="drawer-overlay"></label>
        <ul className="menu p-4 w-80 bg-base-100 text-base-content">
          <li>
            <a>Sidebar Item 1</a>
          </li>
          <li>
            <a>Sidebar Item 2</a>
          </li>
        </ul>
      </div>
    </div>
  );
}
```

**Comment by saadeghi** [comment]
It's really hard to help when you're not providing a reproduction repo.  
I can guess you're handling some part of the styles dynamically.  
- Maybe you're checking/unchecking the checkbox value using JS and the default value is `checked`
- Maybe you're adding some classes dynamically using JS so on the first load, the style is different because JS is not loaded yet.

It can be anything. Please provide an example repo so I can help.

**Reply by cqh963852** [reply]
I have create a codesandbox, But I can't reproduce this problem on codesandbox.

https://codesandbox.io/p/sandbox/modest-forest-n477kw

and Here is my code https://github.com/cqh963852/nextjs-13-daisyui

**Reply by cqh963852** [reply]
It can only be reproduced in rare cases, and I haven't found a sure path

**Comment by cqh963852** [comment]
![image](https://user-images.githubusercontent.com/17702287/232537544-cf4c1d83-c278-40a4-b5a3-19fdc6418e24.png)

![image](https://user-images.githubusercontent.com/17702287/232537288-08ceba00-4aa0-4f44-ab99-1f33600019b9.png)

Is it possible that layout.css has not been loaded yet? flash-of-unstyled-content?

**Comment by cqh963852** [comment]
I have create another repo.The css and html is copy from nextjs's cache. https://github.com/cqh963852/daisyui-test

My way of reproducing

1. npx http-server
2. open browser dev tool, open elements tab
3. access http://localhost:8080

![animate](https://user-images.githubusercontent.com/17702287/232828523-71ba739a-18ac-4b07-a401-aafdd329842e.gif)

![animate2](https://user-images.githubusercontent.com/17702287/232830952-95cd9b02-d08c-43fb-8f55-04c0f3e7b681.gif)

same on firefox


![firefox-animate](https://user-images.githubusercontent.com/17702287/232833625-f68e8d2a-a551-47fc-a38a-efc65edfea73.gif)

**Comment by saadeghi** [comment, marked as answer]
I see...  
Thank you for providing examples and videos.  
This will be fixed in daisyUI version 3.0 which is expected to be released this May. I rewrote the drawer style for version 3.0 and it fixes this issue and many other issues. Unfortunately I can't add a hotfix for this in version 2.x because it will be a breaking change for current apps using daisyUI but you can soon upgrade to version 3.0 when it's released.  

If you're interested you can install `daisyui@0.0.0-insiders.3.5` to try the new style (it's still alpha version, do not use it in production but you can see the drawer issue is fixed)

Let me know if you have any questions.

**Reply by cqh963852** [reply]
Although there is a problem in the reproduction, I don't know what the root cause of this problem is, can you explain it to me? 

I still wonder if it might be caused by the browser's dev tools.

**Reply by saadeghi** [reply]
> I still wonder if it might be caused by the browser's dev tools.

Yes and no!  
The problem is the order of CSS styles. First, the drawer grid is defined to show sidebar and content side by side, then sidebar goes off-canvas using CSS transform. When you open devtools and use `slow 3G` and disable the cache, every thing loads slowly including the CSS file which runs line by line. So the CSS line which puts the sidebar off-canvas, runs after the the lines that show the sidebar in the grid layout.  
This probably won't be visible in a normal speed because with a normal internet connection, CSS loads faster than `slow 3G` simulation and it definitely won't be visible on the 2nd and 3rd load because CSS will be cached.

**Reply by cqh963852** [reply]
Using `daisyui@0.0.0-insiders.3.5` resolve the flash problem. (Focus on dev tool's Elements + Network slow 3G)

---

I read some articles about FOUC https://stackoverflow.com/a/61122747/7275293

What I believe is 

> nothing could be painted on a screen under normal circumstances until the render tree

I have record the proformance and update it to the [repo](https://github.com/cqh963852/daisyui-test).

For the reproduce. The first time `Recalculate Style` only takes 2.07ms (The button's outline is rendered, But color not)
![image](https://user-images.githubusercontent.com/17702287/233138931-46ea2df2-c244-48d7-918e-1b26a66477e8.png)
![image](https://user-images.githubusercontent.com/17702287/233145083-30be349b-b5e9-470b-9a4e-f1d26dd0567c.png)

For the none-reproduce. The first time `Recalculate Style` takes 6.94ms
![image](https://user-images.githubusercontent.com/17702287/233139351-d42729cd-4b4b-481b-90a9-9a1f07c81ecd.png)

I guess it is mainly because the execution process of the browser is different here, resulting in different rendering results.
